### PR TITLE
Added feature for detecting common heads automatically.

### PR DESCRIPTION
**Pull request by JayBaileyCS**
# Description

Added a feature for detecting common heads automatically, by using a detection pattern to elementwise multiply across an attention pattern, only keeping the attention you're interested in. There's support for previous token heads, duplicate token heads, and induction heads. Users can also define their own detection_pattern. (For example, in IOI, you could create a detection pattern with a 1 if a given token's word is capitalised and 0 otherwise, allowing you to detect how much attention is paid to the sentence's names)

The feature is self-contained, so it does not break anything else, and has 100% test coverage.

I have not yet documented this feature aside from a docstring in the code itself - open to ideas for how we might want to go about this.

The code is also probably not as optimised as it could be, certainly open to ideas on that!

## Type of change

- [ x ] New feature (non-breaking change which adds functionality)

# Checklist:

- [ x ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ x ] My changes generate no new warnings
- [ x ] I have added tests that prove my fix is effective or that my feature works
- [ x ] New and existing unit tests pass locally with my changes
- [ x ] I have not rewritten tests relating to key interfaces which would affect backward compatibility.

### CODE REVIEW

**Comment by JayBaileyCS** [comment]
https://colab.research.google.com/drive/1NfCi7DP17gWiX1wT-hDDmIHUpZJaaQOm#scrollTo=aU1ReJigXOsH shows it being briefly used.

File: transformer_lens/detect_heads.py
Comment (jbloomAus) [inline transformer_lens/detect_heads.py:1]: Stuff to add:

- [ ] Docstrings to the get head pattern functions (explain what each is a link if needed, such as to ioi paper)
- [ ] Type hints for return arguments 
- [ ] Example usage (just for detect_head_type) in a form that will be picked up nicely in the docs (see IOI evals dataset example).

File: tests/unit/test_detect_heads.py
Comment (jbloomAus) [inline tests/unit/test_detect_heads.py:1]: Good stuff!

**Review by jbloomAus — CHANGES_REQUESTED** [review]
Thanks Jay, this is a valuable contribution. 

Besides the specific points, I was wondering if you could:

- [x] 1. Ensure this functionality is mentioned where relevant in any of the tutorials and if not add a mini tutorial for this feature.
- [ ] 2. Share a notebook or something where you validate the code aligned with analysis done by others (ie: whatever convinced you that your implementation works). 

Much appreciated!

File: transformer_lens/detect_heads.py
Comment (JayBaileyCS) [reply transformer_lens/detect_heads.py:1]: I have now added all these things. Could I get another look?

**Comment by JayBaileyCS** [comment]
Changes have been added.

File: transformer_lens/detect_heads.py
Comment (jbloomAus) [reply transformer_lens/detect_heads.py:1]: great thanks!

**Review by jbloomAus — APPROVED** [review]

**Comment by jbloomAus** [comment]
Looks great! Thanks Jay!

**Pull request merged**